core/solver/batch_richardson.cpp

```diff
@@ -0,0 +1,134 @@
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/solver/batch_richardson.hpp>
+
+
+#include <ginkgo/core/matrix/batch_dense.hpp>
+#include <ginkgo/core/preconditioner/batch_preconditioner_strings.hpp>
+
+
+#include "core/solver/batch_richardson_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace batch_rich {
+
+
+GKO_REGISTER_OPERATION(apply, batch_rich::apply);
+
+
+}  // namespace batch_rich
+
+
+template <typename ValueType>
+std::unique_ptr<BatchLinOp> BatchRichardson<ValueType>::transpose() const
+{
+    return build()
+        .with_preconditioner(parameters_.preconditioner)
+        .with_max_iterations(parameters_.max_iterations)
+        .with_rel_residual_tol(parameters_.rel_residual_tol)
+        .with_relaxation_factor(parameters_.relaxation_factor)
+        .on(this->get_executor())
+        ->generate(share(
+            as<BatchTransposable>(this->get_system_matrix())->transpose()));
+}
+
+
+template <typename ValueType>
+std::unique_ptr<BatchLinOp> BatchRichardson<ValueType>::conj_transpose() const
+{
+    return build()
+        .with_preconditioner(parameters_.preconditioner)
+        .with_max_iterations(parameters_.max_iterations)
+        .with_rel_residual_tol(parameters_.rel_residual_tol)
+        .with_relaxation_factor(conj(parameters_.relaxation_factor))
+        .on(this->get_executor())
+        ->generate(share(as<BatchTransposable>(this->get_system_matrix())
+                             ->conj_transpose()));
+}
+
+
+template <typename ValueType>
+void BatchRichardson<ValueType>::apply_impl(const BatchLinOp *const b,
+                                            BatchLinOp *const x) const
+{
+    using Vector = matrix::BatchDense<ValueType>;
+    using real_type = remove_complex<ValueType>;
+    auto exec = this->get_executor();
+    auto dense_b = as<const Vector>(b);
+    auto dense_x = as<Vector>(x);
+    const kernels::batch_rich::BatchRichardsonOptions<real_type> opts{
+        parameters_.preconditioner, parameters_.max_iterations,
+        parameters_.rel_residual_tol, parameters_.relaxation_factor};
+
+    log::BatchLogData<ValueType> logdata;
+    // allocate logging arrays assuming uniform size batch
+    const size_type num_rhs = dense_b->get_size().at(0)[1];
+    const size_type num_batches = dense_b->get_num_batch_entries();
+    batch_dim<> sizes(num_batches, dim<2>{1, num_rhs});
+    logdata.res_norms =
+        matrix::BatchDense<real_type>::create(this->get_executor(), sizes);
+    logdata.iter_counts.set_executor(this->get_executor());
+    logdata.iter_counts.resize_and_reset(num_rhs * num_batches);
```

Comment:
The allocation time should be negligible compared to the entire solve. Also, I think it would be best to wipe the logger after one solve before starting the next anyway.